File: app/src/main/java/com/absinthe/anywhere_/adapter/page/PageListAdapter.java
package com.absinthe.anywhere_.adapter.page;

import com.absinthe.anywhere_.R;
import com.absinthe.anywhere_.adapter.ItemTouchCallBack;
import com.chad.library.adapter.base.BaseNodeAdapter;
import com.chad.library.adapter.base.entity.node.BaseNode;

import org.jetbrains.annotations.NotNull;

import java.util.Collections;
import java.util.List;

public class PageListAdapter extends BaseNodeAdapter implements ItemTouchCallBack.OnItemTouchListener {

    public PageListAdapter() {
        addNodeProvider(new PageTitleProvider());
        addNodeProvider(new PageProvider());
        addChildClickViewIds(R.id.iv_entry);
    }

    @Override
    protected int getItemType(@NotNull List<? extends BaseNode> list, int i) {
        BaseNode node = list.get(i);
        if (node instanceof PageTitleNode) {
            return 1;
        } else if (node instanceof PageNode) {
            return 2;
        }
        return -1;
    }

    @Override
    public void onMove(int fromPosition, int toPosition) {
        if (fromPosition < toPosition) {
            for (int i = fromPosition; i < toPosition; i++) {
                Collections.swap(getData(), i, i + 1);
            }
        } else {
            for (int i = fromPosition; i > toPosition; i--) {
                Collections.swap(getData(), i, i - 1);
            }
        }
        notifyItemMoved(fromPosition, toPosition);
    }

    @Override
    public void onSwiped(int position) {

    }
}

File: app/src/main/java/com/absinthe/anywhere_/adapter/page/PageListAdapter.kt
package com.absinthe.anywhere_.adapter.page

import com.absinthe.anywhere_.R
import com.absinthe.anywhere_.adapter.ItemTouchCallBack
import com.chad.library.adapter.base.BaseNodeAdapter
import com.chad.library.adapter.base.entity.node.BaseNode
import java.util.*

class PageListAdapter : BaseNodeAdapter(), ItemTouchCallBack.OnItemTouchListener {

    init {
        addNodeProvider(PageTitleProvider())
        addNodeProvider(PageProvider())
        addChildClickViewIds(R.id.iv_entry)
    }

    override fun getItemType(data: List<BaseNode>, position: Int): Int {
        return when (data[position]) {
            is PageTitleNode -> 1
            is PageNode -> 2
            else -> -1
        }
    }

    override fun onMove(fromPosition: Int, toPosition: Int) {
        if (fromPosition < toPosition) {
            for (i in fromPosition until toPosition) {
                Collections.swap(data, i, i + 1)
            }
        } else {
            for (i in fromPosition downTo toPosition + 1) {
                Collections.swap(data, i, i - 1)
            }
        }
        notifyItemMoved(fromPosition, toPosition)
    }

    override fun onSwiped(position: Int) {}
}